%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% For more detailed article preparation guidelines, please see:
% http://f1000research.com/author-guidelines

\documentclass[10pt,a4paper,twocolumn]{article}
\usepackage{f1000_styles}
\usepackage{listings}
\usepackage{hyperref}

%% Default: numerical citations
\usepackage[numbers]{natbib}

%% Uncomment this lines for superscript citations instead
% \usepackage[super]{natbib}

%% Uncomment these lines for author-year citations instead
% \usepackage[round]{natbib}
% \let\cite\citep

% settings for code blocks (listings pkg)
% see here: https://latex.ugent.be/files/documenten/listings.pdf
\lstset{
  basicstyle=\small,
  xleftmargin=.05\textwidth, xrightmargin=.05\textwidth
}

\begin{document}

\title{Large-scale sequence comparisons with \textit{sourmash}}
%\titlenote{no_note}
\author[1*]{N. Tessa Pierce}
\author[1*]{Luiz Irber}
\author[1,2*]{Taylor Reiter}
\author[1*]{Phillip Brooks}
\author[1**]{C. Titus Brown}
\affil[*]{equal work}
\affil[1]{Department of Population Health and Reproduction, School of Veterinary Medicine, University of California, Davis}
\affil[2]{Food Science Graduate Group, University of California, Davis}
\affil[**]{correspondence: ctbrown@ucdavis.edu}

\maketitle
\thispagestyle{fancy}


\begin{abstract}

The \lstinline{sourmash} software package uses MinHash-based sketching to create “signatures”, compressed representations of DNA, RNA, and protein sequences, that can be stored, searched, explored, and taxonomically annotated. \lstinline{sourmash} signatures can be used to estimate sequence similarity between very large data sets quickly and in low memory, and can be used to search large databases of genomes for matches to query genomes and metagenomes. \lstinline{sourmash} is implemented in C++, Rust, and Python, and is freely available under the BSD license at \lstinline{http://github.com/dib-lab/sourmash}.


\end{abstract}

\section*{Keywords}

bioinformatics, sequence analysis, MinHash, k-mer, sourmash

\clearpage

\section*{Introduction}

Bioinformatic analyses rely on sequence comparison for many applications, including variant analysis, taxonomic classification and functional annotation. As the Sequence Read Archive now contains over 20 Petabases of data \cite{ncbi_2018}, there is great need for methods to quickly and efficiently conduct similarity searches on a massive scale. MinHash techniques \cite{broder1997resemblance} utilize random sampling of k-mer content to generate small subsets known as "sketches" such that Jaccard similarity (the intersection over the union) of two sequence data sets remains approximately equal to their true Jaccard similarity \cite{broder1997resemblance, ondov2016mash}. The many-fold size reductions gained via MinHash opens the door to extremely large scale searches.

While the initial k-mer MinHash implementation focused on enabling Jaccard similarity comparisons \cite{ondov2016mash}, it has since been modified and extended to enable k-mer abundance comparisons \cite{boveefinch}, decrease runtime and memory requirements \cite{zhao2018bindash}, and work on streaming input data \cite{rowe2019streaming}. Furthermore, as Jaccard similarity is impacted by the relative size of the sets being compared, containment searches \cite{broder1997resemblance, koslicki2019improving, mash_screen} have been developed to enable detection of a small set within a larger set, such as a genome within a metagenome.

Here we present version 2.0 of \lstinline{sourmash} \cite{brown2016sourmash}, a Python library for building and utilizing MinHash sketches of DNA, RNA, and protein data. \lstinline{sourmash} incorporates and extends standard MinHash techniques for sequencing data, with a particular focus towards enabling efficient containment queries using large databases. This is accomplished with two modifications: (1) building sketches via a modulo approach \cite{broder1997resemblance}, and (2) implementing a modified Sequence Bloom Tree \cite{solomon2016fast} to enable both similarity and containment searches. In most cases, these features enable \lstinline{sourmash} database comparisons in sub-linear time.

Standard genomic MinHash techniques, first implemented in Ondov BD \textit{et al.} \cite{ondov2016mash}, retain the minimum \textit{n} k-mer hashes as a representative subset. \lstinline{sourmash} extends these methods by incorporating a user-defined "scaled" factor to build \lstinline{sourmash} sketches via a modulo approach, rather than the standard bottom-hash approach \cite{broder1997resemblance}. Sketches built with this approach retain the same fraction, rather than number, of k-mer hashes, compressing both large and small datasets at the same rate. 

This enables comparisons between datasets of disparate sizes but can sacrifice some of the memory and storage benefits of standard MinHash techniques, as the signature size scales with the number of unique k-mers rather than remaining fixed \cite{mash_screen}. In \lstinline{sourmash}, use of the "scaled" factor enables user modification of the trade-off between search precision and sketch size, with the caveat that searches and comparisons can only be conducted using signatures generated with identical "scaled" values (downsampled at the same rate). 

To enable large-scale database searches using these signatures, \lstinline{sourmash} implements a modified Sequence Bloom Tree (SBT), the SBT-MinHash (SBTMH), that allows both similarity (\lstinline{sourmash search}) and containment (\lstinline{sourmash gather}) searches for taxonomic exploration and classification. Notably, Jaccard similarity searches using this modified SBT require storage of the cardinality of the smallest MinHash below each node in order to properly bound similarity. \lstinline{sourmash} also implements a second database format, "LCA", for in-memory search when sufficient RAM is available or database size is tractable. The LCA format can be leveraged to return additional information, such as taxonomic lineage.

In addition to these modifications, \lstinline{sourmash} implements k-mer abundance tracking \cite{boveefinch} within signatures to allow abundance comparisons across datasets and facilitate metagenome, metatranscriptome, and transcriptome analyses, and is compatible with streaming approaches. 
The \lstinline{sourmash} library is implemented in C++, Rust \cite{Matsakis:2014:RL:2692956.2663188}, and Python, and can be accessed both via command line and Python API. The code is available under the BSD license at \lstinline{http://github.com/dib-lab/sourmash}.


\section*{Implementation}

\lstinline{sourmash} provides a user-friendly, extensible platform for MinHash signature generation and manipulation for DNA, RNA, and protein data. Sourmash is designed to facilitate containment queries for taxonomic exploration and identification while maintaining all functionality available via standard genomic MinHash techniques. 

\subsection*{\lstinline{sourmash} Signatures}

\lstinline{sourmash} modifies standard genomic MinHash techniques in two ways. First, \lstinline{sourmash} scales the number of retained hashes to better represent and compare datasets of varying size and complexity. Second, \lstinline{sourmash} optionally tracks the abundance of each retained hash, to better represent data of metagenomic and transcriptomic origin and allow abundance comparisons.

\textit{\textbf{Scaling}} \lstinline{sourmash} implements a method inspired by modulo sketches \cite{broder1997resemblance} to dynamically scale hash subset retention size (\textit{n}). When using scaled signatures, users provide  a scaling factor (\textit{s}) that divides the hash space into \textit{s} equal bands, retaining hashes within the minimum band as the sketch. These scaled signatures can be converted to standard bottom-hash signatures, if the subset retention size \textit{n} is equal to or smaller than the number of hashes in the scaled signature. \lstinline{sourmash} provides a signature utility, \lstinline{downsample}, to convert sketches when possible. Finally, to maintain compatibility with sketches generated by other programs such as Mash \cite{ondov2016mash}, \lstinline{sourmash} generates standard bottom-hash MinHash sketches if users specify the hash subset size \textit{n} rather than scaling factor.


\textit{\textbf{Streaming compatibility}} Scaled signature generation is streaming compatible and provides some advantages over streaming calculation using standard MinHash. As data streams in, standard MinHash replaces hashes based on the minimum hash value to maintain a fixed number of hashes in the signature. In contrast, no hash is ever removed from a scaled signature as more data is received. As a result, for searches of a database using streamed-in data, all prior matches remain valid (although their significance may change as more data is received). This allows us to place algorithmic guarantees on containment searches using streaming data.


\textit{\textbf{Abundance tracking}} \lstinline{sourmash} extends MinHash functionality by implementing abundance tracking of k-mers. k-mer counts are incremented after hashing as each k-mer is added to the hash table. \lstinline{sourmash} tracks abundance for k-mers in the minimum band and stores this information in the signature. These values accompany the hashes in downstream comparison processes, making signatures better representations of repetitive sequences of metagenomic and transcriptomic origin. 

 
\textit{\textbf{Signatures}} MinHash sketches associated with a single sequence file are stored together in a “signature” file, which forms the basis of all \lstinline{sourmash} comparisons. Signatures may include sketches generated with different \textit{k} sizes or molecule type (nucleotide or protein) and are stored in JSON format to maintain human readability and ensure proper interoperability. 

Signatures can only be compared against signatures and databases made from the same parameters (\textit{k} size(s), scaled value, nucleotide or protein level). If signatures differ in their scaled value or size(\textit{n}), the larger signatures can be downsampled to become comparable with smaller signatures using the \textit{signature utilities}, below. \lstinline{sourmash} also provides 6-frame nucleotide translation to generate protein signatures from nucleotide input if desired.

\textit{\textbf{Signature utilities}} \lstinline{sourmash} provides a number of utilities to facilitate set operations between signatures (\lstinline{merge}, \lstinline{intersect}, \lstinline{extract}, \lstinline{downsample}, \lstinline{flatten} \lstinline{subtract}, \lstinline{overlap}), and handling (\lstinline{describe}, \lstinline{rename},  \lstinline{import}, \lstinline{export}) of \lstinline{sourmash} signatures. These can be accessed via the \lstinline{sourmash signature} subcommand. 

\subsection*{SBT-MinHash}
\lstinline{sourmash} implements a modified Sequence Bloom Tree (SBT \cite{solomon2016fast}), the SBT-MinHash (SBTMH), which can efficiently capture large volumes of MinHashes (e.g., all microbes in GenBank) and support multiple search regimes that improve on run time of linear searches. 

\textit{\textbf{Implementation}} The SBTMH is a n-ary tree (binary by default), where leaf nodes are MinHash signatures and internal nodes are Bloom Filters. Each Bloom Filter contains all the values from its children, so the root node contains all the values from all signatures. SBTMH is designed to be extensible such that signatures can be subsequently added without the need for full regeneration. Adding a new signature to SBTMH causes parent nodes up to the root to be updated, but other nodes are not affected.

SBTMH trees can be combined if desired: In the simplest case, adding a new root and updating it with the content of the previous roots is sufficient, and this preserves all node information without changes. As an example, separate indices can be created for each RefSeq subdivision (bacteria, archaea, fungi, etc) and be combined depending on the application (such as an analysis for bacteria + archaea, but not fungi). In practice, this is most useful for updating the SBTMH, as both \lstinline{search} and  \lstinline{gather} support search over multiple databases without the need for rebuilding a single large database.

\textit{\textbf{Searching SBTMH}} Similarity searches start at the root of the SBTMH, and check for query elements present in each internal node. If the similarity does not reach the threshold, the subtree under that node does not need to be searched. If a leaf is reached, it is returned as a match to the query signature. In order to enable similarity (in addition to containment) searches using this modified SBTMH, nodes store the cardinality of the smallest signature below each node in order to properly bound similarity. The full SBTMH does not need to be imported to RAM during searches, making this method best for rapid searching with minimal memory requirements. However, if sufficient RAM is available, searches of databases (or many signatures) may be completed in memory via an alternate database format (discussed below). 

\textit{\textbf{SBTMH utilities}} \lstinline{sourmash} provides several utilities for construction, use, and handling of SBTMH databases. These include \lstinline{sbt index} to index a collection of signatures as an SBTMH for fast searching, \lstinline{sbt append} to add signatures, and \lstinline{sbt combine} to join two SBTMH databases.


\subsection*{LCA database}

\lstinline{sourmash} implements an alternate database format, LCA, to support in-memory queries. This implementation utilizes two named lists to store MinHashed databases: the first containing MinHashes, and the second containing taxonomic information, with both lists named by sample name. This structure facilitates direct look-up of MinHashes, and thus can be leveraged to return additional information, such as taxonomic lineage. The LCA database structure can be prepared using the \lstinline{sourmash lca index} command.

\section*{Assessing sequence similarity}

\subsection*{Pairwise comparisons}

For sequence comparison, \lstinline{sourmash compare} re-implements Jaccard sequence similarity comparison to enable comparison between scaled MinHashes. When abundance tracking of k-mers is enabled, \lstinline{compare} instead calculates the cosine similarity, although we recommend using more accurate approaches for detailed comparisons \cite{rowe2019streaming}.

\subsection*{Database searches}
In addition to conducting pairwise comparisons, two types of database searches are implemented: breadth-first similarity searches (\lstinline{sourmash search}) and best-first containment searches (\lstinline{sourmash gather}), which support different biological queries. These searches can be conducted using either database format. 

\textit{\textbf{Similarity queries}} Breadth-first \lstinline{sourmash search} can be used to obtain all MinHashes in the SBTMH that are present in the query signature (above a specified threshold). This style of search is streaming-compatible, as the query MinHash can be augmented as the search is occurring.


\textit{\textbf{Containment queries}}
 Best-first \lstinline{sourmash gather} implements a greedy algorithm where the SBTMH is descended on a linear path through a set of internal nodes until the highest containment leaf is reached. The hashes in this leaf are then subtracted from the query MinHash and the process is repeated until the threshold minimum is reached. \lstinline{sourmash} post-processes similarity statistics after the search such that it reports percent identity and unique identity for each match.

\textit{\textbf{Taxonomy-resolved searches}}
\lstinline{sourmash} can conduct taxonomy-resolved searches uses the “least common ancestor” approach (as in Kraken \cite{wood2014kraken}), to identify k-mers in a query. From this it can either find a consensus taxonomy between all the k-mers (\lstinline{sourmash classify}) or it can summarize the mixture of k-mers present in one or more signatures (\lstinline{sourmash summarize}).

\subsection*{Operation}

\lstinline{sourmash} is a tool for building and utilizing MinHash signatures of DNA, RNA, and protein sequences. A straightforward workflow consists of generating a signature using \lstinline{sourmash compute}, and comparing it against other signatures or databases of signatures via \lstinline{sourmash compare, search, gather, lca search,} or \lstinline{lca gather}. \lstinline{sourmash} has no particular memory requirements, but does need to hold the largest single sequence in memory while generating a signature. For example, computing a signature from a ~100Mb human microbiome sample requires ~30MB of RAM, and searching it against a sourmash Genbank signature database takes 1-6 minutes and requires 2-6 GB of RAM, depending on the search type. "LCA" databases are smaller on disk but require more memory to be searched.

Below we provide several use cases to demonstrate the utility of \lstinline{sourmash} for sequence comparisons, starting with signature generation and proceeding into signature comparisons, tetranucleotide frequency clustering analysis, and taxonomic classification. We primarily demonstrate nucleotide-level applications in this paper; protein-level analyses will be explored further in future work. Additional information and tutorials are available at \url{https://sourmash.readthedocs.io}. 


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Use cases}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\subsection*{Installation}

\lstinline{sourmash} is available for both Linux and OSX, and runs under either \lstinline{Python 2.7.x} or \lstinline{Python 3.5+}. To install \lstinline{sourmash}, we recommend using \href{https://docs.conda.io/projects/conda/en/latest/}{conda}. For these examples, we used sourmash v.2.0.1, installed with conda v 4.6.14.

\begin{lstlisting}
conda install -c conda-forge \
-c bioconda sourmash
\end{lstlisting}

Alternate installation instructions are available at \url{sourmash.readthedocs.io}. 

\subsection*{Creating a signature}

All \lstinline{sourmash} comparisons work on signatures, compressed representations of biological sequencing data. To create a signature from sequences with abundance tracking:

\begin{lstlisting}
# download the genome
curl -L  https://osf.io/bjh2y/download \
-o GCF_000005845.2_ASM584v2_genomic.fna.gz

# calculate the signature
sourmash compute -k 21,31,51 \ 
  --scaled 2000 \ 
  --track-abundance \ 
  -o GCF_000005845.2_ASM584v2_genomic.sig \
  GCF_000005845.2_ASM584v2_genomic.fna.gz
\end{lstlisting}

Because a signature can contain multiple MinHashes, multiple k-sizes can be specified per a sequence. Only one scaled size can be used. 

By default, the name of the file becomes the name of the signature. To name the signature from the first line of the sequencing file, use \lstinline{--name-from-first}. Although the \lstinline{--track-abundance} flag is optional, since downstream comparison methods contain the flag \lstinline{--ignore-abundance} to ignore them, we recommend calculating all signatures with abundance tracking. 

\smallskip

To create a signature from protein sequences:

\begin{lstlisting}
# download amino acid sequences
curl -L https://osf.io/y9kra/download \ 
-o GCF_000146045.2_R64_protein.faa.gz

# calculate the signature
sourmash compute -k 11,21,31 \
    --scaled 2000 \
    --track-abundance \
    -o GCF_000146045.2_R64_protein.sig \ 
    GCF_000146045.2_R64_protein.faa.gz
\end{lstlisting}

Signatures can also be made directly from reads. Depending on the downstream use cases, we recommend different preparation methods. When the user aims to \lstinline{compare} the signature to other signatures, we recommend k-mer trimming the reads before computing the signature. Because \lstinline{compare} does an all-by-all comparison of signatures, errors in the reads will falsely deflate the similarity metric. We recommend trimming RNA-seq or metagenome reads with \lstinline{trim-low-abund.py} in the \href{https://pypi.org/project/khmer/}{khmer package} \cite{crusoe2015khmer}, a dependency of \lstinline{sourmash}. 

\begin{lstlisting}
# download the reads
curl -L -o ERR458584.fq.gz \ 
    https://osf.io/pfxth/download

# trim the reads
trim-low-abund.py ERR458584.fq.gz \ 
    -V -Z 10 -C 3 --gzip -M 3e9 \ 
    -o  ERR458584.khmer.fq.gz 

# calculate the signature from trimmed reads
sourmash compute -k 21,31,51 \ 
    --scaled 2000 \ 
    --track-abundance \ 
    -o ERR458584.khmer.sig \ 
    ERR458584.khmer.fq.gz
\end{lstlisting}

When using methods that compare a signature against a database such as \lstinline{gather} or \lstinline{search}, k-mer trimming need not be used. These methods use exact matching of hashes in the signature to those in the databases. k-mer trimming could increase false negatives, but results on k-mer trimmed data will more accurately represent the proportions of content in the data.

\begin{lstlisting}
# calculate the signature from raw reads
sourmash compute -k 21,31,51 \ 
    --scaled 2000 \ 
    --track-abundance \ 
    -o ERR458584.sig \ 
    ERR458584.fq.gz
\end{lstlisting}

\subsection*{Comparing many signatures}

Signatures calculated with abundance tracking enable rapid comparison of sequences where k-mer frequency is variable, and can be leveraged for quality control and summarization methods. For example, principle component analysis (PCA) and multidimensional scaling (MDS) are standard quality control and summarization methods for count data generated during RNA-seq analysis \cite{conesa2016survey}. \lstinline{sourmash} can be used to build this MDS plot in a reference-free (or assembly-free) manner, using k-mer abundances of the input reads. We also find this useful for comparing other types of RNA sequencing samples (mRNA, ribo-depleted, 3' tag-seq, metatranscriptomes, and transcriptomes). 

\subsubsection*{MDS}

Here, we use a set of four \textit{Saccharomyces cerevisiae} RNA-seq samples: replicate wild-type samples and replicate mutant (\textit{SNF2}) samples \cite{schurch2016many}. To use \lstinline{sourmash} to build an MDS plot, we first trim the data to remove low abundance k-mers via khmer \cite{crusoe2015khmer}. We demonstrate the streaming capability of \lstinline{sourmash} by downloading, k-mer trimming, and calculating a signature with one command. This allows the user to generate signatures without needing to store large files locally.

\begin{lstlisting}
curl -L https://osf.io/pfxth/download \ 
  | trim-low-abund.py -V -Z 10 \ 
  -C 3 -M 3e9 -o - -\ 
  | sourmash compute -k 31 \ 
  --scaled 2000 --track-abundance \ 
  -o ERR458584.khmer.sig -
\end{lstlisting}

The signature will be named from the input filename, in this case \lstinline{-}. We can change the name to reflect its contents using the \lstinline{signature rename} function. 

\begin{lstlisting}
sourmash signature rename \ 
    -k 31 -o ERR458584.khmer-named.sig \ 
    ERR458584.khmer.sig \ 
    ERR458584.khmer 
\end{lstlisting}

We can also check that the name has been changed. 

\begin{lstlisting}
sourmash signature describe \ 
    ERR458584.khmer-named.sig
\end{lstlisting}

Using signatures from four samples, we can compare the files with the \lstinline{compare} function. Here we download signatures calculated and renamed using the above commands. We output the comparison matrix as a csv for downstream use in other platforms.

\begin{lstlisting}
# download signatures
curl -L -o yeast-sigs.tar.gz \ 
    https://osf.io/pk2w5/download

# uncompress the signatures
tar xvf yeast-sigs.tar.gz

# compare the signatures
sourmash compare -k 31 \ 
  --csv yeast-comp.csv \ 
  *named.sig
\end{lstlisting}

We then import the \lstinline{compare} similarity matrix into R (v3.4.1) to produce an MDS plot with wild-type samples (ERR459011, ERR459102) in yellow and mutant samples (ERR458584, ERR458829) in blue.

\begin{lstlisting}
# Read data into R
comp_mat <- read.csv("yeast-comp.csv")

# Set row labels
rownames(comp_mat) <- colnames(comp_mat)

# Transform for plotting
comp_mat <- as.matrix(comp_mat)

# Make an MDS plot
fit <- dist(comp_mat)
fit <- cmdscale(fit)
plot(fit[ , 2] ~ fit[ , 1], 
     xlab = "Dim 1", 
     ylab = "Dim 2",
     xlim= c(-.6, .9),
     main = "sourmash Compare MDS")
# add labels to the plot
text(fit[ , 2]~ fit[ , 1], 
     labels = row.names(fit), 
     pos = 4, font = 1,
     data = fit, 
     col = c("blue", "blue", 
             "orange", "orange"))
\end{lstlisting}

For comparison, we also produced an MDS plot using a more traditional approach, utilizing \href{https://github.com/COMBINE-lab/salmon}{Salmon} (v0.11.3) \cite{patro2017salmon} to quantify abundance relative to an \textit{S. cerevisiae} reference, and \href{https://bioconductor.org/packages/release/bioc/html/edgeR.html}{edgeR} (v3.22.5) \cite{robinson2010edger} to build an MDS plot (code available online at https://osf.io/97rt4/). 


\begin{figure}
\centering
\includegraphics[width=0.4\textwidth]{sourmash_comparison.pdf}
\caption{\label{fig:sourmash_comparison} The MDS plots produced from the reference-free \lstinline{sourmash compare} similarity matrix and the transcript quantification analysis (salmon and edgeR) are similar. Wild-type \textit{S. cerevisiae} samples (ERR459011, ERR459102) are in yellow and mutant samples (ERR458584, ERR458829) in blue.}
\end{figure}



\subsubsection*{Tetranucleotide Frequency Clustering}
We can also use \lstinline{sourmash} with abundance tracking for tetranucleotide frequency clustering. Tetranucleotide usage is species-specific, with strongest conservation in DNA coding regions \cite{Pride2003}. This is often used in metagenomics as one method to "bin" assembled contiguous sequences together that are from the same species \cite{Albertsen2013}. Recently, tetranucleotide frequency clustering using \lstinline{sourmash} was used to detect microbial contamination in the domesticated olive genome \cite{Reiter2018}. Here we reimplement this approach using 100 of the 11,038 scaffolds in the draft genome. We calculate the signature using a k-mer size of 4, use all 4-mers, and track abundance. Then we use \lstinline{sourmash compare} to calculate the similarity between each scaffold. (The \lstinline{--singleton} flag, calculates a signature for each sequence in the fasta file.)

\begin{lstlisting}
# download the subsampled genome
curl -L https://osf.io/xusfa/download \
    -o Oe6.scaffolds.sub.fa.gz 
    
# calculate a signature for each scaffold
sourmash compute -k 4 \ 
    --scaled 1 \ 
    --track-abundance \ 
    --singleton \ 
    -o Oe6.scaffolds.sub.sig \ 
    Oe6.scaffolds.sub.fa.gz 

# pairwise compare all scaffolds
sourmash compare -k 4 \ 
    -o Oe6.scaffolds.sub.comp \ 
    Oe6.scaffolds.sub.sig
\end{lstlisting}

Although \lstinline{sourmash compare} supports export to a csv file, \lstinline{sourmash} also has a built in visualization function, \lstinline{plot}. We will use this to visualize scaffold similarity.

\begin{lstlisting}
sourmash plot --labels \ 
    --vmin .4 \ 
    Oe6.scaffolds.sub.comp 
\end{lstlisting}

In Figure 2, we see that there is high similarity between 98 of the scaffolds, but that Oe6\_s01156 and Oe6\_s01003 are outliers with tetranucleotide frequency similarity around 40\% to olive scaffolds. These two scaffolds are contaminants \cite{Reiter2018}. 

\begin{figure}
\centering
\includegraphics[width=0.4\textwidth]{olive_genome_heatmap.pdf}
\caption{\label{fig:olive_genome_heatmap} Heatmap and dendrogram generated using \lstinline{sourmash} signatures built from scaffolds in the domesticated olive genome. Two scaffolds are outliers when using tetranucleotide frequency to calculate similarity (highlighted in green on the dendrogram).}
\end{figure}

\subsection*{Comparisons to detect outliers}

MinHash comparisons are useful for outlier detection. Below we compare 50 genomes that contain the word "\textit{Escherichia coli}." We have pre-calculated the signatures for each of these genomes. We then use the \lstinline{plot} function to visualize our comparison.

\begin{lstlisting}
# download the signatures into 
# a folder
mkdir escherichia-sigs
cd escherichia-sigs

curl -L https://osf.io/pc76j/download \ 
    -o escherichia-sigs.tar.gz    

# decompress the signatures
tar xzf escherichia-sigs.tar.gz
rm escherichia-sigs.tar.gz

cd ..

# pairwise compare the signatures
sourmash compare -k 31 \ 
    -o ecoli.comp \ 
    escherichia-sigs/*sig
    
# plot the comparison
sourmash plot --labels \ 
    ecoli.comp
\end{lstlisting}

\begin{figure}
\centering
\includegraphics[width=0.4\textwidth]{compmatrix.pdf}
\caption{\label{fig:compmatrix} Heatmap and dendrogram generated using \lstinline{sourmash} signatures built from 50 genomes that contained the word "\textit{Escherichia coli}." One signature is an outlier (highlighted in blue on the dendrogram).}
\end{figure}

We see that the minimum similarity in the matrix is 0\%. If all 50 signatures were from the same species, we would expect to observe higher minimum similarity at a k-mer size of 31. When we look closely, we see one signature has 0\% similarity with all other signatures because it is a phage (Figure 3).

\subsection*{Classifying signatures}

The \lstinline{search} and \lstinline{gather} functions allow the user to classify the contents of a signature by comparing it to a database of signatures. Prepared databases for microbial genomes in RefSeq and GenBank are available at \url{https://sourmash.readthedocs.io/en/latest/databases.html}. However, it is also simple to create a custom database with signatures. 

\bigskip

Below we make a database that contains 50 \textit{Escherichia coli} genomes.


\begin{lstlisting}
mkdir escherichia-sigs
cd escherichia-sigs

curl -L https://osf.io/pc76j/download \ 
    -o escherichia-sigs.tar.gz    

tar xzf escherichia-sigs.tar.gz
rm escherichia-sigs.tar.gz

cd ..

sourmash index -k 31 ecolidb \ 
    escherichia-sigs/*.sig
\end{lstlisting}

\smallskip

This database can be queried with \lstinline{search} and \lstinline{gather} using any signature calculated with a k-size of 31.


For example, below we download a small set of k-mer trimmed \textit{Escherichia coli} reads and generate a signature with k=31.

\smallskip

\begin{lstlisting}
curl -L -o ecoli-reads.khmer.fq.gz \ 
    https://osf.io/26xm9/download \ 
    
sourmash compute -k 31 
    --scaled 2000 \ 
    ecoli-reads.khmer.fq.gz \  
    -o ecoli-reads.sig
\end{lstlisting}

Then, we search the 50-genome database created above.

\lstinline{sourmash search}: 

\begin{lstlisting}
sourmash search -k 31 \
    ecoli-reads.sig ecolidb \  
    --containment
\end{lstlisting}

\begin{lstlisting}[basicstyle=\tiny]
49 matches; showing first 3:
similarity   match
----------   -----
 65.4%       NZ_JMGW01000001.1 Escherichia coli 1-176-05_S4_C2 e117605...
 64.9%       NZ_GG774190.1 Escherichia coli MS 196-1 Scfld2538, whole ...
 63.7%       NZ_JMGU01000001.1 Escherichia coli 2-011-08_S3_C2 e201108...
\end{lstlisting}

Breadth-first \lstinline{sourmash search} finds all matches in the SBTMH that are present in the query signature (above a specified threshold).

Now try the same search using \lstinline{sourmash gather}.

\begin{lstlisting}
sourmash gather -k 31 \ 
    ecoli-reads.sig ecolidb
\end{lstlisting}

\begin{lstlisting}[basicstyle=\tiny,]
loaded query: ecoli_ref-5m.khmer.fq.gz... (k=31, DNA)
loaded 1 databases.

overlap     p_query p_match
---------   ------- -------
4.1 Mbp       65.4%   83.5%    NZ_JMGW01000001.1 Escherichia coli 1-...
2.4 Mbp        2.7%    3.3%    NZ_GG749254.1 Escherichia coli FVEC14...
3.4 Mbp        1.4%    1.7%    NZ_MOGK01000001.1 Escherichia coli st...
3.1 Mbp        0.6%    0.7%    NZ_LVOV01000001.1 Escherichia coli st...
3.1 Mbp        0.3%    0.4%    NZ_MIWP01000001.1 Escherichia coli st...
3.0 Mbp        0.3%    0.4%    NZ_APWY01000001.1 Escherichia coli 17...
3.5 Mbp        0.2%    0.2%    NZ_JNLZ01000001.1 Escherichia coli 3-...
4.0 Mbp        0.2%    0.2%    NZ_GG774190.1 Escherichia coli MS 196...
2.3 Mbp        0.1%    0.2%    NZ_KB732756.1 Escherichia coli KTE66 ...
2.0 Mbp        0.1%    0.1%    NZ_BBUW01000001.1 Escherichia coli O1...
2.3 Mbp        0.1%    0.1%    NZ_MOZX01000101.1 Escherichia coli st...
2.3 Mbp        0.1%    0.1%    NZ_JSMW01000001.1 Escherichia coli st...
4.0 Mbp        0.1%    0.1%    NZ_JMGU01000001.1 Escherichia coli 2-...
2.0 Mbp        0.0%    0.0%    NZ_MOZC01000010.1 Escherichia coli st...
2.1 Mbp        0.0%    0.0%    NZ_MKJG01000001.1 Escherichia coli st...
1.8 Mbp        0.0%    0.0%    NZ_AEKA01000453.1 Escherichia sp. TW1...
2.6 Mbp        0.0%    0.0%    NZ_LEAD01000071.1 Escherichia coli st...
3.5 Mbp        0.0%    0.0%    NZ_MIWF01000001.1 Escherichia coli st...

found 18 matches total;
the recovered matches hit 71.5% of the query
\end{lstlisting}

Best-first \lstinline{sourmash gather} finds the best match first, e.g. here the first \textit{E. coli} genome has an 83\% match to 65.4\% of our query signature. The hashes that matched (65.4\% of the query) are then subtracted, and the database is queried with the remaining hashes (34.6\% of original query). This process is repeated until the threshold is reached. \lstinline{sourmash} post-processes similarity statistics after the search such that it reports percent identity and unique identity for each match.

\lstinline{sourmash gather} is also useful for rapid metagenome decomposition. Below we calculate a signature of a metagenome using raw reads, and then use \lstinline{gather} to perform a best-first search against all microbial genomes in Genbank. This approach was recently used to classify unknown genomes in a "mock" metagenome \cite{Shakya2013}. The mock community was made to contain 64 genomes, but additional genomic material was inadvertently added prior to sequencing. Below we will use \lstinline{gather} to investigate the content in the mock metagenome that did not map to the 64 reference genomes. For details on how this signature was created, please see Awad \textit{et al.}\cite{Awad2017}. Note that the GenBank database is approximately 7.8 Gb compressed, and 50 Gb decompressed. Searches of the current GenBank database run fastest if allowed to use ~11 Gb of RAM.

\begin{lstlisting}
# download the signature
curl -L -o unmapped-qc-to-ref.fq.sig\ 
    https://osf.io/ /download \ 

# download the gather k 31 Genbank database
curl -L -o genbank-d2-k31.tar.gz \ 
    https://s3-us-west-2.amazonaws.com/
    sourmash-databases/2018-03-29/ 
    genbank-d2-k31.tar.gz
    
# run gather
sourmash gather -k 31 \ 
    -o unmapped-qc-to-ref.csv \ 
    unmapped-qc-to-ref.fq.sig \ 
    genbank-d2-k31
\end{lstlisting}

The output to the terminal begins:

\begin{lstlisting}[basicstyle=\tiny,]
loaded query: unmapped-qc-to-ref.fq... (k=31, DNA)
downsampling query from scaled=10000 to 10000
loaded 1 databases.


overlap     p_query p_match
---------   ------- -------
1.6 Mbp        1.1%   19.9%    BA000019.2 Nostoc sp. PCC 7120 DNA, c...
1.2 Mbp        0.8%   50.8%    LN831027.1 Fusobacterium nucleatum su...
1.2 Mbp        0.8%   31.0%    CP001957.1 Haloferax volcanii DS2 pla...
1.1 Mbp        0.8%   16.7%    BX119912.1 Rhodopirellula baltica SH ...
1.0 Mbp        0.7%   29.0%    CH959311.1 Sulfitobacter sp. EE-36 sc...
0.8 Mbp        0.6%   37.3%    AP008226.1 Thermus thermophilus HB8 g...
0.8 Mbp        0.6%   56.7%    CP001941.1 Aciduliprofundum boonei T4...
0.8 Mbp        0.5%   23.3%    FOVK01000036.1 Proteiniclasticum rumi...
0.7 Mbp        0.5%   15.0%    CP000031.2 Ruegeria pomeroyi DSS-3, c...
0.7 Mbp        0.5%   11.3%    CP000875.1 Herpetosiphon aurantiacus ...
0.6 Mbp        0.4%   22.9%    BA000023.2 Sulfolobus tokodaii str. 7...
0.6 Mbp        0.4%   13.6%    AP009153.1 Gemmatimonas aurantiaca T-...
\end{lstlisting}

We see that 20.1\% of k-mers match 82 genomes in GenBank. The majority of matches are to genomes present in the mock community. However, some species like \textit{Proteiniclasticum ruminis} were not members of the mock community. These results also highlight how \lstinline{sourmash gather} behaves with inexact matches such as strain variants. For example, we see two matches between \textit{P. ruminis} strains among all matches. This likely indicates that a \textit{P. ruminis} strain that has not been sequenced before is in our sample, and that it shares more k-mers of size 31 in common with one strain than the other. (See Brown CT \textit{et al.} \cite{spacegraphcats} for further analysis of this strain.)

\lstinline{sourmash gather} and \lstinline{search} also support custom databases. Using a custom database with \lstinline{sourmash gather}, we can idnetify the dominant contamination in the domesticated olive genome \cite{Reiter2018}. Below, we will use a database containing all fungal genomes in NCBI. We will then use the streaming compatibility of \lstinline{sourmash} to download and calculate the signature. Lastly, we will search the olive genome against the fungal genomes using \lstinline{gather}.

\begin{lstlisting}
# download the fungal database
curl -L -o fungi-genomic.tar.gz \ 
    https://osf.io/7yzc4/download

# decompress the database
tar xf fungi-genomic.tar.gz 

# download the olive genome 
# calculate the signature
curl -L https://osf.io/k9358/download \ 
  | zcat \ 
  | sourmash compute -k 31 \ 
  --scaled 2000 --track-abundance \ 
  -o Oe6.scaffolds.sig -
 
# perform gather
sourmash gather -k 31 \ 
    --scaled 2000 \ 
    -o Oe6.scaffolds.csv \
    Oe6.scaffolds.sig \ 
    fungi-k31
\end{lstlisting}

Using \lstinline{gather}, we see two matches both within the genus \textit{Aureobasidium}. This is the dominant contaminant within the genome \cite{Reiter2018}.

\begin{lstlisting}[basicstyle=\tiny,]
loaded query: Oe6.scaffolds.fa... (k=31, DNA)
loaded 1 databases.


overlap     p_query p_match avg_abund
---------   ------- ------- ---------
140.0 kbp      0.0%    1.0%       1.2    LVWM01000001.1 Aureobasidium pullulan...
68.0 kbp       0.0%    0.1%       1.0    MSDY01000045.1 Aureobasidium sp. FSWF...
found less than 30.0 kbp in common. => exiting

found 2 matches total;
the recovered matches hit 0.0% of the query
\end{lstlisting}


\section*{Conclusions}

The \lstinline{sourmash} package provides a collection of tools to conduct sequence comparisons and taxonomic classification, and makes comparison against large-scale databases such as GenBank and SRA tractable on laptops. \lstinline{sourmash} signatures are small and irreversible, which means they can be used to facilitate pre-publication data sharing that may help improve classification databases and facilitate comparisons among similar datasets. 

\subsection*{Data availability}
\emph{Underlying data} 

Open Science Framework: sourmash-use-cases (DOI: 10.17605/OSF.IO/KESH2)
This project contains the following underlying data:
\begin{itemize}
\item data-files
      \begin{itemize}
          \item ecoli-reads.khmer.fq.gz (Small set of k-mer trimmed \textit{Escherichia coli} reads)
          \item ERR458584.fq.gz (\textit{Saccharomyces cerevisiae} SRA Record ERR458584 SNF2 mutant, \cite{schurch2016many})
          \item Oe6.scaffolds.fa.gz (domesticated olive (\textit{Olea europaea}) genome \cite{Reiter2018})
          \item Oe6.scaffolds.sub.fa.gz (Subsampled set of scaffolds from the domesticated olive (\textit{Olea europaea}) genome \cite{Reiter2018})
          \item yeast\_ktrimmed.tar (kmer-trimmed \textit{Saccharomyces cerevisiae} reads  \cite{schurch2016many})
      \end{itemize}
\item index-files
      \begin{itemize}
          \item escherichia-sigs.tar.gz (Sourmash signatures of 50 randomly selected \textit{Escherichia coli} genomes)
          \item fungi-genomic.tar.gz (Sourmash signature database of all fungal genomes in NCBI as of 12/2018)
      \end{itemize}
\item signature-files
      \begin{itemize}
          \item GCF\_000005845.2\_ASM584v2\_genomic.fna.gz (\textit{Escherichia coli} genome str. K-12 substr. MG1655)
          \item GCF\_000146045.2\_R64\_genomic.fna.gz (\textit{Saccharomyces cerevisiae} S288C genome)
          \item GCF\_000146045.2\_R64\_protein.faa.gz (\textit{Saccharomyces cerevisiae} S288C protein sequence)
      \end{itemize}
\end{itemize}

\emph{Extended data}

Open Science Framework: sourmash-use-cases (10.17605/OSF.IO/KESH2)
This project contains the following extended data:

\begin{itemize}
    \item yeast-mds.txt (Code to generate MDS plots via Salmon and edgeR)
\end{itemize}

Licence: CC0 1.0 Universal

\subsection*{Software availability}

Source code available from: \url{https://github.com/dib-lab/sourmash/}\\
Archived source code at time of publication: DOI: 10.5281/zenodo.3240653\\
Licence: \href{https://opensource.org/licenses/BSD-3-Clause}{3-Clause BSD License} \\

\subsection*{Competing interests}
The authors declare no competing interests.

\subsection*{Grant information}
This work is funded in part by the Gordon and Betty Moore Foundation’s Data-Driven Discovery Initiative [GBMF4551 to CTB]. NTP was supported by an National Science Foundation Postdoctoral Fellowship in Biology [1711984].

%\subsection*{Acknowledgements}

{\small\bibliographystyle{unsrtnat}
\bibliography{sourmash}}

\end{document}
